text/0000-javascript-module-api.md

```diff
@@ -343,6 +343,24 @@ Export alternative](#everything-is-a-named-export).
 
 [decorators]: http://tc39.github.io/proposal-decorators/
 
+### One Level Deep
+
+To avoid deep directory hierarchies with mostly-empty directories, this proposal
+limits nesting inside a top-level package to a single level. Deep nesting like
+this can add additional time to navigating the hierarchy without adding much
+benefit.
+
+Java packages often have this problem due to their URL-based namespacing; see
+e.g. [this Java
+library](https://github.com/elvishew/xLog/tree/fbfb60f9472e32723436b3d6bdd6c1878a5afb37/library/src)
+where you end up with deeply nested directories, like
+`xLog/library/src/test/java/com/elvishew/xlog/printer/AndroidPrinterTest.java`.
+
+This rule leads to including the type in the name of the module in some cases
+where it might otherwise be grouped instead. For example, instead of
+`@ember/routing/locations/none`, we prefer `@ember/routing/none-location` to
+avoid the second level of nesting.
```

Comment:
I think the rule of one level deep is in conflict with two other stated goals: [discoverability](https://github.com/emberjs/rfcs/pull/176#issuecomment-259066709) and [Organize by mental model](https://github.com/tomdale/rfcs/blob/js-modules/text/0000-javascript-module-api.md#organize-by-mental-model). Also, the comparison with Java seems like a straw man. 

**Discoverability & mental model**: `@ember/routing/locations/none` allows a developer to intuit that there are other routing location possbilities. Since `locations` is like a folder, I can intuit there are other things inside that folder. By similar logic, I can look at the parent, and intuit that `routing` has a few packages inside it -- `locations` being just one of them. The path lets me know I can go to `ember/routing` and discover those packages and increase my understanding of the routing module. If I step into `ember/routing/locations`, I can discover the other options for locations.

By comparison, `@ember/routing/none-location`, makes the capabilites of the routing module entirely opaque. It doesn't indicate there are other routing location options nor indicate how the routing package is organized or what it's capable of. Instead of searching for `ember/routing/locations` or navigating in the source code to `ember/routing/locations`, I'd go to `ember/routing` and have to filter through all the other modules inside the routing package. Further, it's not intuitive that I need to reverse "none" and "location" to come up with the import path of "routing/none-location". In my mind, this last problem is the worst.

The multi level path describes my mental model of Ember. The single level path doesn't and forces me to have one mental model for how the framework works and one for how Ember authors convert from the framework to import paths. With multi level paths, learning what to import helps teach what's inside ember and what ember can do. With single level paths, learning what to import becomes a gigantic memorization game.

**Regarding Java**, the example you gave is **10!!** path segments long when all the Ember examples in the RFC are 3 paths max. We're not even close to having the complexity of the example and it's inclusion here feels like a poor comparison that doesn't prove the point you're trying to make.

**My suggestion:** 
To step back for a moment, the point isn't the locations example nor is it to argue for super deep paths. I agree -- deeply nested paths are awful. I think there's a middle ground between arbitrarily long paths and one level deep. If we can find that middle ground, I think we can avoid deeply nested paths while also supporting the goals of discoverability and mapping the imports to a developers mental model. To fix the dilema, I'd suggest the rule be changed to become a few suggestions:
- prefer single level trees
- choose clarity over brevity: use longer paths only if it will help with discoverability and/or better conform to Ember's mental model